cmd/thanos/compact.go

```diff
@@ -268,15 +268,33 @@ func runCompact(
 
 	if retentionByResolution[compact.ResolutionLevelRaw].Seconds() != 0 {
 		level.Info(logger).Log("msg", "retention policy of raw samples is enabled", "duration", retentionByResolution[compact.ResolutionLevelRaw])
+		if retentionByResolution[compact.ResolutionLevelRaw].Seconds() <= downsample.DownsampleRange0 {
+			level.Warn(logger).Log("msg", "retention policy of raw samples is too low to generate 5 min aggregated samples", "duration", retentionByResolution[compact.ResolutionLevelRaw])
+		}
 	}
 	if retentionByResolution[compact.ResolutionLevel5m].Seconds() != 0 {
 		level.Info(logger).Log("msg", "retention policy of 5 min aggregated samples is enabled", "duration", retentionByResolution[compact.ResolutionLevel5m])
+		if retentionByResolution[compact.ResolutionLevel5m].Seconds() <= downsample.DownsampleRange1 {
+			level.Warn(logger).Log("msg", "retention policy of 5 min aggregated samples is too low to generate 1h aggregated samples", "duration", retentionByResolution[compact.ResolutionLevel5m])
+		}
 	}
 	if retentionByResolution[compact.ResolutionLevel1h].Seconds() != 0 {
 		level.Info(logger).Log("msg", "retention policy of 1 hour aggregated samples is enabled", "duration", retentionByResolution[compact.ResolutionLevel1h])
 	}
 
 	f := func() error {
+		// At first, we need at least to be able to genetes 1h resolution metrics
+		// from the lower resolutions, so remove only up to 1h resolution retention.
```

Comment:
I feel like there is an implicit assumption here that retention times are non-decreasing and "correct". What if the user sets the 1h retention to some very small value? Won't we delete some data that we are not supposed to? At least right now we would generate it but, of course, we would delete it immediately :smile: 

Also, I thought we were going to only apply this at the beginning if the retention times are configured to be the same as otherwise we run the risk of losing data. Is there something missing as to why you are avoiding that check? :smile: 

How did you test this out?

Reply:
Nice catch. I'll add the cases of non-decreasing somewhere, and add tests as well.